### PR TITLE
Dhiraj/active vision

### PR DESCRIPTION
**Pull request by Dhiraj100892**
# Description

Data collection using slam and label propagation using robot trajectory data
* Made changes to PyRobot to continuously collect data in habitat and build the corresponding docker image `theh1ghwayman/locobot-assistant:6.0`
* Added label propagation code using robot trajectory and depth data. 

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

If this PR introduces a change or fixes a bug, please add before and after screenshots (if this is causes a UX change) or before and after examples(this could be plain text, videos etc ) to demonstrate the change.

If this PR introduces a new feature, please also add examples or screenshots demonstrating the feature.

# Testing

Tested locally as well as on robot

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [x] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

File: locobot/label_propogation/label_propogation_parallel.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    """
+    R = Rotation.from_euler("Z", current_pose[2]).as_matrix()
+    XYZ = np.matmul(XYZ.reshape(-1, 3), R.T).reshape((-1, 3))
+    XYZ[:, 0] = XYZ[:, 0] + current_pose[0]
+    XYZ[:, 1] = XYZ[:, 1] + current_pose[1]
+    return XYZ
+
+
+@ray.remote
+def propogate_label(
```
Comment (anuragprat1k) [inline locobot/label_propogation/label_propogation_parallel.py:45]: does this use the notion of skip frames mentioned in the readme or is that somewhere else?

File: locobot/label_propogation/label_propogation_parallel.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                    if (
+                        min(np.linalg.norm(cur_pts_in_world[loc] - gt_pix_depth_in_world, axis=1))
+                        > dist_thr
+                    ):
+                        filtered_img_indx[pixel_index] = False
+
+            # take out filtered pix values
+            pts_in_cur_img = pts_in_cur_img[filtered_img_indx]
+
+            # step to take care of quantization erros
```
Comment (anuragprat1k) [inline locobot/label_propogation/label_propogation_parallel.py:198]: what do you mean by quantization errors? add a note

File: locobot/label_propogation/label_propogation_parallel.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+@ray.remote
+def propogate_label(
+    root_path: str,
+    src_img_indx: int,
+    src_label: np.ndarray,
+    propogation_step: int,
+    base_pose_data: np.ndarray,
+    out_dir: str,
+):
+    """Take the label for src_img_indx and propogate it to [src_img_indx - propogation_step, src_img_indx + propogation_step]
```
Comment (anuragprat1k) [inline locobot/label_propogation/label_propogation_parallel.py:53]: add more details here - a textual explanation of the approach basically, something like a pseudo code. Does this need depth for the entire trajectory? Also, what is the idea behind the depth filtering?

File: locobot/robot/slam_pkg/README.md [hunk 4]
```diff
@@ -13,14 +13,17 @@ roscore &
...
+python slam.py --robot habitat --goal 10 2 0 --map_size 4000 --robot_rad 25 --vis --save_vis --store_path ./tmp --dataset_path /Replica-Dataset --scene apartment_0
 ```
 
  Results should look something like this
 <p align="center">
     <img src="https://media.giphy.com/media/eTof1wrCZHXw83tGwH/giphy.gif", width="960" height="192">
 </p>
- 
+
+- `save_vis` arg will also store `rgb, depth, seg, robot trajectory(data.json)` under `store_path(args)/scene(args)` folder
```
Comment (anuragprat1k) [inline locobot/robot/slam_pkg/README.md:24]: how are `--vis` and `--save_vis` different?

File: locobot/robot/slam_pkg/slam.py [hunk 5]
```diff
@@ -531,39 +600,85 @@ def main(args):
...
+                "scene_path": "{}/{}/habitat/mesh_semantic.ply".format(args.dataset_path, scene),
+            }
+            args.store_path = "./tmp/{}".format(scene)
+            robot = Robot("habitat", common_config=config)
+            agent_state = robot.base.agent.get_state()
+            # place the robot at place where it can move
+            p = robot.base.sim.pathfinder.get_random_navigable_point()
+
+            agent_state.position = copy(p)
+            agent_state.sensor_states["rgb"].position = copy(p + np.array([0.0, 0.6, 0.0]))
```
Comment (anuragprat1k) [inline locobot/robot/slam_pkg/slam.py:617]: add a note here about the `[0, 0.6, 0]` values.

File: locobot/robot/slam_pkg/README.md [hunk 6]
```diff
@@ -13,14 +13,17 @@ roscore &
 
 ```
```
Comment (anuragprat1k) [inline locobot/robot/slam_pkg/README.md:14]: can you also add details about the slam implementation in the readme here? Add references to papers /  implementations where necessary. How does it work? What are its gaps? What can be the future improvements? etc.

**Review by anuragprat1k — CHANGES_REQUESTED** [review]
Thanks for putting this out. Requesting more documentation changes.

File: .circleci/habitat_docker/Dockerfile [hunk 7]
```diff
@@ -56,8 +56,7 @@ RUN patchelf --set-rpath '$ORIGIN' /root/pyenv_pyrobot_python3/lib/python3.6/sit
 
 # tested pyrobot commit that works with Habitat
 RUN cd ~/low_cost_ws/src/pyrobot && git checkout Develop && \
-    git pull https://github.com/facebookresearch/pyrobot Develop && \
-    git checkout de3e54ea9e1196ea96caf8672646d5dfd63a3731 && \
+    git pull https://github.com/Dhiraj100892/pyrobot.git cont_move_habitat && \
```
[the remark this answers is not in the document]
Comment (Dhiraj100892) [reply .circleci/habitat_docker/Dockerfile:59]: yes, I can raise the PR for it

File: .circleci/habitat_docker/locobot_assistant.dockerfile [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # syntax = docker/dockerfile:experimental
 # FROM nvidia/cudagl:10.1-base-ubuntu18.04 as base
-FROM dhirajgandhi/pyrobot-and-habitat:1.0 as base
+FROM dhirajgandhi/pyrobot-and-habitat:1.1 as base
```
[the remark this answers is not in the document]
Comment (Dhiraj100892) [reply .circleci/habitat_docker/locobot_assistant.dockerfile:3]: yes

File: locobot/label_propogation/create_coco_data.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    else:
+        with open(train_json, "r") as f:
+            train_json_data = json.load(f)
+            train_json_data = train_json_data["img_id"]
+    img_id_set = set()
+    if test:
+        # for testing include all the images except the which were as gt during label prpopogation
+        # NOTE: for testing provide img_annot_root_path which has gt annotation
+        for i in range(len(glob.glob(os.path.join(img_root_path, "*.jpg")))):
+            if i not in train_json_data:
```
[the remark this answers is not in the document]
Comment (Dhiraj100892) [reply locobot/label_propogation/create_coco_data.py:80]: its not the best assumption. I did it based on the current data collection pipeline which always try to create between `[0, num of images)`

File: locobot/robot/slam_pkg/README.md (hunk 4 above)
Comment (Dhiraj100892) [reply locobot/robot/slam_pkg/README.md:24]: `--vis` will show plt window while robot is exploring, this is helpful when there is screen attached to robot
`--save_vis` will store exploration visualization in addition to rgbd & seg data

**Pull request closed without merging**